Refactor models tasks usecases

mahshidaln opened the pull request:
# PR Type ([Feature | Fix | Documentation | Test])
Fix

## Short Description
Apply changes to models, tasks, and use-cases:
- models: remove multiple dispatch, update SKModel params, refactor methods
- tasks: remove multiple dispatch, follow the models update
- use-cases: update the entire mortality notebook, fix CXR notebook

## Tests Added
...

amrit110 reviewed (commented):
@mahshidaln  Could you also remove the dependencies from project? You can do `poetry remove multipledispatch` and `poetry remove evaluate`

mahshidaln commented:
@fcogidi I made a small change to the evaluator: I removed the batch_size from the list of the `predict` method parameters and added it to the model config file for SKModel.

fcogidi commented on cyclops/models/wrappers/pt_model.py:
```diff
@@ -881,94 +903,96 @@ def predict_proba(self, X, **predict_params):
+                for feature in feature_columns:
+                    stacked.append(torch.stack(examples[feature]))
+                X = torch.cat(stacked, dim=1)
+                return {pred_column: self.predict(X)}
```
Instead of doing a recursive call here, I propose calling `_evaluation_step` directly in addition to applying the activation function. That is:
```python
output = self._evaluation_step(batch, training=False, **predict_params)
output = self.activation_(output)
```
That way, a dataloader will not be created for every call to `predict_proba`.

fcogidi commented on cyclops/models/wrappers/pt_model.py:
```diff
@@ -729,63 +728,86 @@ def partial_fit(self, X, y=None, **fit_params):
+            ):
+                self.partial_fit(X)
+        else:
+            if y is None:
```
`y` could remain optional for unsupervised models.

fcogidi commented on cyclops/models/wrappers/pt_model.py:
```diff
@@ -729,63 +728,86 @@ def partial_fit(self, X, y=None, **fit_params):
+                columns=feature_columns + target_columns,
+                **format_kwargs,
+            ):
+                self.partial_fit(X)
```
How to get features and target in the training loop?

fcogidi commented on cyclops/models/wrappers/sk_model.py:
```diff
@@ -47,10 +56,22 @@ class SKModel:
+        model: SKBaseEstimator,
+        model_params: dict,
+        fit_params: dict = {},
+        best_model_params: dict = {},
```
Why not pass `best_model_params` and `fit_params` during hyperparameter tuning and model fitting, respectively?

mahshidaln replied:
Because they were already in the config file. Also I wanted to remove batch_size from the methods parameters so I thought similar to PTModel, add have everything set up from the beginning. You think it's not a good idea?

fcogidi replied:
Yeah. I think the difference in initialization style is consistent with the workflow in the two frameworks.

fcogidi commented on cyclops/models/wrappers/sk_model.py:
```diff
@@ -81,66 +102,142 @@ def initialize_model(self, **kwargs):
         return self
 
-    def find_best(
+    def find_best(  # pylint: disable=too-many-branches
```
Note that there is a [base class](https://github.com/VectorInstitute/cyclops/blob/7f30a685e432e856b2d987b7329248e16220e456/cyclops/models/wrappers/base.py) that is intended to serve as a template for implementing model wrappers (this one didn't inherit it because the methods were added dynamically), and it has this signature for the `find_best` method:
```python
def find_best(
        self,
        X,
        y,
        parameters: Union[Dict, List[Dict]],
        metric: Optional[Union[str, Callable, Sequence, Dict]] = None,
        method: Literal["grid", "random"] = "grid",
        **kwargs,
    )
 ```
 I think the `SKModel` should follow the same signature. Otherwise, the signature for the base and PTModel wrappers should be updated.
 
 Ditto for the other methods.

amrit110 commented:
@mahshidaln I noticed this weird failure with psycopg2 with the delirium pipeline as well. To fix it you can freeze the version using "poetry add psycopg2==2.9.5".
